### PR TITLE
Developer guide docs

### PR DESCRIPTION
**Pull request by strangetom**
This PR will add some details developer guidance documentation. See #861 and #617 for discussion.

I plan to cover the following:

- [x] Introductory docs
- [x] Step by step guide
- [ ] In depth: debugging during development
- [x] In depth: HTML scraping (best practice, common patterns etc.)
- [x] In depth: Ingredient groups
- [x] In depth: Scraper functions (all possible function, which are optional, mandatory etc., what their content should contain, return types, ...)

It will take me a bit of time to write all these, but I want to open this draft PR to get feedback as I do it. I do expect this to cause some discussion as the aim is to document things that have not necessarily been written down before.

All opinions and contributions are welcome.

### CODE REVIEW

File: docs/README.md
Comment (jayaddison) [inline docs/README.md:11]: +1 @strangetom - this is a nice way of explaining the goals.  I agree with placing accuracy and authenticity in the same line item, too - I was thinking about exactly that during a commute recently and figured that those two reduce to more-or-less the same idea.

File: docs/how-to-develop-scraper.md
Comment (jayaddison) [inline docs/how-to-develop-scraper.md:7]: We might want to link to the fully-qualified README URL ( https://github.com/hhursev/recipe-scrapers/blob/main/README.rst ) here - because if these docs are built and deployed somewhere in future, that relative URL might not work (most of the others should.. but this one is outside the `docs` directory)

File: docs/in-depth-guide-ingredient-groups.md
Comment (jayaddison) [inline docs/in-depth-guide-ingredient-groups.md:5]:
```suggestion
Sometimes a website will format lists of ingredients using groups, where each group contains the ingredients needed for a particular aspect of the recipe. Recipe schema has no way to represent these groupings, so all of the ingredients are presented as a single list and information about the groupings are lost.
```

File: docs/in-depth-guide-ingredient-groups.md
Comment (jayaddison) [inline docs/in-depth-guide-ingredient-groups.md:44]: Trying to make some of the wording here a bit more direct.  I don't think my suggestions are perfect, so I think some more edits may be needed:

```suggestion
Adding ingredient group support to a scraper involves overriding the `ingredient_group` function for it.  There are three important points to consider:

1. The schema.org Recipe format does not support groupings - so scraping HTML is going to be required in the implementation. 
2. The `group_ingredients` function requires that the schema.org ingredients list matches the ingredients found from the HTML - so it's important to make sure that these match.
3. Not all recipes on a website will use ingredient groups, so the implementation must degrade gracefully in cases where groupings aren't available. For recipes that don't have ingredient groups, the output should be the same as default implementation (i.e. a single `IngredientGroup` with `purpose=None` and `ingredients=ingredients()`).
```

File: docs/in-depth-guide-ingredient-groups.md
Comment (jayaddison) [inline docs/in-depth-guide-ingredient-groups.md:57]: Not directly related to this documentation: I've wondered before whether we should relocate `group_ingredients` onto `AbstractScraper` so that we don't have to pass the `soup` argument here.

Maybe if-and-when that is refactored, that'd be a good case for updating documentation and codebase at the same time.

File: docs/in-depth-guide-ingredient-groups.md
Comment (jayaddison) [inline docs/in-depth-guide-ingredient-groups.md:110]:
```suggestion
In these cases supporting ingredient groups may still be possible. The `group_ingredients()` helper method is only that: an optional helper -- you can always implement custom grouping logic yourself by overriding `ingredient_groups()` directly in your scraper if you can't find suitable CSS selectors for the ingredients and groups.
```

**Comment by jayaddison** [comment]
I'm looking forward to the HTML scraping best practices section when that's ready :)

One thing that I'll mention from experience: in web browsers -- and I admit `BeautifulSoup` isn't a browser, but it does end up parsing HTML into element trees in similar ways -- there can be significant performance differences based on how elements are found using CSS selectors  It's almost like SQL database queries: sometimes it's possible to rewrite a query in a way that produces identical results but in a much more efficient and faster way.

Performance isn't top of our priority list, and I think that's fine, but giving readers an awareness that each CSS query _could_ result in processing a really large number of HTML elements (especially given the layout of some websites) -- could be nice.

I'm not sure I have detailed advice there other than something along the lines of: "use `#id` selectors when possible because they may allow unique index lookup, don't be afraid to filter by both class-and-element -- but also try to make your query concise so that when you read it again in six months, or have to adjust it based on a website layout change, it makes some logical sense to you".

**Comment by jayaddison** [comment]
> use #id selectors when possible because they may allow unique index lookup

On second thought: I'm not completely sure whether `lxml` provides this optimization, to be honest.  I'll try to check that in the documentation/code.

**Comment by jayaddison** [comment]
> > use #id selectors when possible because they may allow unique index lookup
> 
> On second thought: I'm not completely sure whether `lxml` provides this optimization, to be honest. I'll try to check that in the documentation/code.

Much of this is irrelevant in our case at the moment, though; we _tend_ to use `BeautifulSoup` operations like `find` and `find_all` that don't depend on CSS selectors.

The few cases where we use CSS selection (usually via `soup.select`) are implemented by `lxml` using [`SoupSieve`](https://github.com/facelessuser/soupsieve.git/), I think.  Note that `lxml` can also use the `cssselect` library to implement some CSS selection operations.

As far as I can tell, `SoupSieve` doesn't build an index of elements-by-ID, so I don't think that using an ID selector (`#foo`) would make a difference there.  In the case of `cssselect` (that I don't believe we use, currently), element-by-ID queries are [transformed into XPath attribute equality checks](https://github.com/scrapy/cssselect/blob/ddd9784977fca6e7da4439d837e4f510f1f10638/cssselect/xpath.py#L395) - so it depends on the underlying `etree`/XPath implementation (native Python, in most cases, probably) how efficient those queries will be.

tl;dr - I don't think my advice around using CSS identity selectors is at all relevant, please disregard that.

(maybe some day we'll have a [large enough dataset](https://www.commoncrawl.org/) to [examine](https://github.com/openculinary/tardir.git/) and improve performance and efficiency of the scrapers, but today it's tricky)

File: docs/README.md
Comment (strangetom) [reply docs/README.md:11]: This was very much inspired by comment you made on a recent PR :)

File: docs/how-to-develop-scraper.md
Comment (strangetom) [reply docs/how-to-develop-scraper.md:7]: Good catch

File: docs/in-depth-guide-ingredient-groups.md
Comment (strangetom) [reply docs/in-depth-guide-ingredient-groups.md:44]: I don't agree with the change to the 2nd point. It isn't that the `group_ingredients` function needs them to match (in fact we added functionally to handle the cases where there isn't an exact match between the two sets of ingredients), it's more than the scraper should be presenting the same thing in two different ways: a single list or in groupings.

File: docs/in-depth-guide-ingredient-groups.md
Comment (strangetom) [reply docs/in-depth-guide-ingredient-groups.md:57]: I had thought along on the same lines when implementing `group_ingredients`. It felt like adding a lot to the `AbstractScraper` that most scrapers probably wouldn't use.

Comment (jayaddison) [reply docs/in-depth-guide-ingredient-groups.md:57]: Yep, seems fair, and it's a small detail.  We can skip that and maybe reconsider it later :+1:

File: docs/README.md
Comment (jayaddison) [reply docs/README.md:11]: Thanks!  I hope they're useful guidelines - I don't want to have too much influence, though! :smile:

@hhursev @bfcarpio @brett as some of the more frequent committers here, do these more-or-less match the aims you have in mind when writing scrapers?  And/or any suggestions for adjustments?

File: docs/in-depth-guide-ingredient-groups.md
Comment (strangetom) [reply docs/in-depth-guide-ingredient-groups.md]: That was original my thinking when writing this. 

However we added a test the `ScraperTest` base class that checks that both methods return the same ingredients, so the wording here should reflect that. I'll have a bit more of a think on how the phrase this.

File: docs/in-depth-guide-scraper-functions.md
Comment (jayaddison) [inline docs/in-depth-guide-scraper-functions.md:97]: +1 on this note also

File: docs/in-depth-guide-scraper-functions.md
Comment (jayaddison) [inline docs/in-depth-guide-scraper-functions.md:118]: The `int` type does have the benefit of being straightforward to emit and consume -- and in some cases it's a direct typecast of the value that was presented on the HTML page.

On the other hand, I suppose [`timedelta`](https://docs.python.org/3/library/datetime.html#datetime.timedelta) could be a more structured alternative?

(I'm not sure I have strong opinions about this - maybe I should have, but I don't.  the only strong opinion I have is that if we do change it, we should migrate all three `*_time` fields during the `v15` migration)

File: docs/in-depth-guide-scraper-functions.md
Comment (jayaddison) [inline docs/in-depth-guide-scraper-functions.md:319]: Nice catch :+1: 

I wonder.. I think we could check how many scrapers _might_ support [`schema.org` reviews](https://schema.org/Review) if we added it as a schema field.  That could take us from one to many fairly quickly.

I'll try to do that analysis soon (note: if I haven't responded to this within a day, then it probably means I've forgotten about it).

Comment (jayaddison) [reply docs/in-depth-guide-scraper-functions.md:319]: > I wonder.. I think we could check how many scrapers _might_ support [`schema.org` reviews](https://schema.org/Review) if we added it as a schema field. That could take us from one to many fairly quickly.

I had an idea for a way to run a naive, basic analysis.  Based on the commands below, 25 of our scrapers look like they contain `Review` objects as structured data:

```shell
$ grep -rwl '@type.*"Review"' tests/test_data | wc -l
25
```

...so maybe 10% of scrpers _could_ return `reviews` if we implement that there?  That would seem valuable enough to open a feature request for.

File: docs/in-depth-guide-scraper-functions.md
Comment (jayaddison) [inline docs/in-depth-guide-scraper-functions.md:252]:
```suggestion
Semi-structured field that can contain a mix of cuisine type (for example, country names), mealtime (breakfast/dinner/etc) and dietary properties (gluten-free, vegetarian).
```

File: docs/in-depth-guide-scraper-functions.md
Comment (strangetom) [reply docs/in-depth-guide-scraper-functions.md:118]: I wasn't sure if a recipe would ever specify a time to sub-minute resolution, in which case we would want to use `float`. 
`int` is probably fine, so I'll leave it like this.

There might be a job to do to check how well the existing scrapers we have align with the return types I've documented here.

File: docs/in-depth-guide-scraper-functions.md
Comment (strangetom) [reply docs/in-depth-guide-scraper-functions.md:319]: If we can extract this from the schema, that I think this is would be worth doing. :+1:

File: docs/in-depth-guide-scraper-functions.md
Comment (strangetom) [reply docs/in-depth-guide-scraper-functions.md:252]: Generally agree, although the scrapers have a `cuisine()` function for cuisine information. That being said, I would expect `category()` to return whatever category the site defines, so that may well include a cuisine.

**Comment by strangetom** [comment]
I've added my first pass at the HTML scraping guide. It turned out to be harder to write than I first thought because I wasn't sure what to actually add to it. It will need some further tweaking, but if anybody has some suggestion I would appreciate them.

File: docs/in-depth-guide-scraper-functions.md
Comment (jayaddison) [reply docs/in-depth-guide-scraper-functions.md:319]: Great - I've filed #870 to track that, and will take a look into more of the details soon.

File: docs/in-depth-guide-html-scraping.md
Comment (jayaddison) [inline docs/in-depth-guide-html-scraping.md:152]: The HTML scraping guide is looking pretty good to me already :)

I'd suggest calling this section "Finding elements using string pattern matching"

Comment (jayaddison) [reply docs/in-depth-guide-html-scraping.md:152]: ...or attribute pattern matching, I suppose (I see your point now that the matching is specific to HTML attributes)

File: docs/in-depth-guide-html-scraping.md
Comment (jayaddison) [inline docs/in-depth-guide-html-scraping.md:9]: Nitpick: typo here: `BeauitfulSoup -> BeautifulSoup`

**Comment by jayaddison** [comment]
Not something I do as often as I should, but an idea for something we could suggest in the developer guide: doing a manual test of the scraper on another recipe from the same website after developing it.

To be honest that's probably also something we should do routinely during code review (again, I admit to not often doing that.  sometimes I do).

**Comment by strangetom** [comment]
I'm marking this as ready for review and merge. I've not written the debugging guide yet, but I don't think I've going to have time in the near future to get to it, so I think it's better to get what is written merged and then add it in the future.

File: README.rst
Comment (jayaddison) [inline README.rst:370]: I think a `blob` URL (like the second hyperlink) is the correct one to go for.  A `tree` URL is to an open branch (branch name `developer-guide-docs`) in this case, which may no longer exist / could be cleaned up after the guide is merged.

Comment (strangetom) [reply README.rst]: Good catch. 
Relative URLs might have been simpler instead of trying to guess what the URL will be once merged, but I think this willl work.

**Review by jknndy — APPROVED** [review]
Pushed a commit with a bunch of small adjustments to spelling, grammar and some syntax consistency changes but looks great to me! 

Thanks for all the time spent on this it is hugely helpful (wish it was around when I started contributing!)

**Comment by jayaddison** [comment]
Brilliant, thanks both - I would generally agree that it makes sense to merge what we have here so far and incrementally update from there.  Merging in a moment :+1: 

(@hhursev: would these be publishable using [GitHub pages](https://docs.github.com/en/pages/getting-started-with-github-pages/configuring-a-publishing-source-for-your-github-pages-site)?)

**Pull request merged**
